Make `Fetcher` Protocol generic on the media type

## Summary

Parameterize the `Fetcher` Protocol on the media type it operates against
so that subclass-specific fetchers (currently `TVMaze`, future `Music`,
`Movie`, etc.) can declare narrower `fetch` signatures without breaking
type-check.

## Motivation

`mypy nielsen` reports one error today:

```
nielsen/processor.py:75: error: Argument 2 to "ProcessorFactory" has
incompatible type "type[TVMaze]"; expected "type[Fetcher]"  [arg-type]
```

The cause is parameter contravariance. `Fetcher` is declared as:

```python
class Fetcher(Protocol):
    def fetch(self, media: nielsen.media.Media) -> None: ...
```

But `TVMaze.fetch` accepts only `TV`, not any `Media`. Because protocol
satisfaction requires that an implementation accept *at least as wide* a
type as the protocol's declaration, `TVMaze` does **not** satisfy
`Fetcher` — the type checker is correctly flagging that passing a
generic `Media` to `TVMaze.fetch` would be a type error.

Suppressing this with `# type: ignore` would hide a real shape mismatch.
The proper fix is to acknowledge in the Protocol that fetchers operate
on a *specific* media subtype.

## Proposed change

Use Python 3.12 generic class syntax (already required by `requires-python
= ">=3.12"`):

```python
# nielsen/fetcher.py
class Fetcher[M: nielsen.media.Media](Protocol):
    def fetch(self, media: M) -> None: ...
```

`TVMaze` then implicitly satisfies `Fetcher[TV]`. Propagate the type
parameter through `Processor` and `ProcessorFactory` in
`nielsen/processor.py`:

```python
@dataclass
class Processor[M: nielsen.media.Media]:
    media_type: type[M]
    fetcher: nielsen.fetcher.Fetcher[M]

@dataclass
class ProcessorFactory[M: nielsen.media.Media]:
    media_type: type[M]
    fetcher: type[nielsen.fetcher.Fetcher[M]]
    def __call__(self) -> Processor[M]: ...
```

The `PROCESSOR_FACTORIES` dict gets `ProcessorFactory[Any]` as its value
type since it's heterogeneous across media types.

## Acceptance criteria

- [ ] `uv run mypy nielsen` reports zero errors.
- [ ] `uv run pytest` continues to pass on 3.12 / 3.13 / 3.14.
- [ ] No `# type: ignore` introduced in `nielsen/fetcher.py` or
      `nielsen/processor.py`.
- [ ] Adding a hypothetical second fetcher (e.g. for a different media
      subclass) would type-check without further protocol changes.

## Out of scope

- Annotating downstream callers of `Processor` / `Fetcher` to take
  advantage of the new generic parameter — that can happen organically
  when each caller is touched.
- Resolving the unrelated `mypy` errors in the test suite (`tests/test_processor.py`,
  `tests/test_config.py`).

--- a/nielsen/fetcher.py
+++ b/nielsen/fetcher.py
@@ -15,11 +15,11 @@ logger: logging.Logger = logging.getLogger(__name__)
 logger.addHandler(logging.NullHandler())
 
 
-class Fetcher(Protocol):
+class Fetcher[M: nielsen.media.Media](Protocol):
     """Used to fetch metadata from an external source rather than infering it from the
     file name."""
 
-    def fetch(self, media: nielsen.media.Media) -> None:
+    def fetch(self, media: M) -> None:
         """Fetch and update metadata using information from the given `Media` object."""
         ...
 

--- a/nielsen/processor.py
+++ b/nielsen/processor.py
@@ -6,13 +6,14 @@ instance and manually calling all of their individual methods, Processors provid
 convenience method (process) which creates a Media instance of the desired type and
 handles all changes to that Media instance."""
 
+import pathlib
 from dataclasses import dataclass
 from enum import Enum
-import pathlib
+from typing import Any
 
-from nielsen.config import config
 import nielsen.fetcher
 import nielsen.media
+from nielsen.config import config
 
 
 class MediaType(str, Enum):
@@ -20,16 +21,16 @@ class MediaType(str, Enum):
 
 
 @dataclass
-class Processor:
+class Processor[M: nielsen.media.Media]:
     """A convenience class to handle all the Media and Fetcher operations required to
     manage an individual file. The `media_type` is a Media class and is used to
     construct instances of the given type. The `fetcher` is an instance of an object
     implementing the `Fetcher` Protocol."""
 
-    media_type: type[nielsen.media.Media]
-    fetcher: nielsen.fetcher.Fetcher
+    media_type: type[M]
+    fetcher: nielsen.fetcher.Fetcher[M]
 
-    def process(self, path: pathlib.Path) -> nielsen.media.Media:
+    def process(self, path: pathlib.Path) -> M:
         """Convenience function to process Media for storage in its library, respecting the configuration
         options (so some steps may be skipped).
         1. Infer metadata based on the Media type.
@@ -40,7 +41,7 @@ class Processor:
         """
 
         # Create an instance of the appropriate Media subclass
-        media: nielsen.media.Media = self.media_type(path)
+        media: M = self.media_type(path)
 
         media.infer()
 
@@ -57,20 +58,20 @@ class Processor:
 
 
 @dataclass
-class ProcessorFactory:
+class ProcessorFactory[M: nielsen.media.Media]:
     """A Factory class for Processors. Given a Media and Fetcher class, calling an
     instance of this Factory will return a Processor with the given Media class and
     Fetcher instance."""
 
-    media_type: type[nielsen.media.Media]
-    fetcher: type[nielsen.fetcher.Fetcher]
+    media_type: type[M]
+    fetcher: type[nielsen.fetcher.Fetcher[M]]
 
-    def __call__(self) -> Processor:
+    def __call__(self) -> Processor[M]:
         """Return a Processor for the given types."""
 
         return Processor(self.media_type, self.fetcher())
 
 
-PROCESSOR_FACTORIES: dict[MediaType, ProcessorFactory] = {
+PROCESSOR_FACTORIES: dict[MediaType, ProcessorFactory[Any]] = {
     MediaType.TV: ProcessorFactory(nielsen.media.TV, nielsen.fetcher.TVMaze),
 }

